components/layout_2020/display_list/mod.rs

```diff
@@ -124,7 +100,105 @@ impl Fragment {
                     wr::ColorF::WHITE,
                 );
             },
+            Fragment::Text(t) => {
+                self.build_display_list_for_text_fragment(t, builder, containing_block)
+            },
+        }
+    }
+
+    fn build_display_list_for_text_fragment(
+        &self,
+        fragment: &TextFragment,
+        builder: &mut DisplayListBuilder,
+        containing_block: &PhysicalRect<Length>,
+    ) {
+        // NB: The order of painting text components (CSS Text Decoration Module Level 3) is:
+        // shadows, underline, overline, text, text-emphasis, and then line-through.
+
+        builder.is_contentful = true;
+
+        let rect = fragment
+            .rect
+            .to_physical(fragment.parent_style.writing_mode, containing_block)
+            .translate(containing_block.origin.to_vector());
+        let mut baseline_origin = rect.origin.clone();
+        baseline_origin.y += fragment.font_metrics.ascent;
+        let glyphs = glyphs(&fragment.glyphs, baseline_origin);
+        if glyphs.is_empty() {
+            return;
+        }
+
+        let mut common = builder.common_properties(rect.to_webrender());
+        common.hit_info = hit_info(&fragment.parent_style, fragment.tag, Cursor::Text);
+
+        let color = fragment.parent_style.clone_color();
+        let text_decorations = fragment
+            .parent_style
+            .get_inherited_text()
+            .text_decorations_in_effect;
```

Comment:
It would be good to not use the `text_decorations_in_effect` mechanism, because it's pretty busted, see https://searchfox.org/mozilla-central/rev/d5b34cc8872177d3ee071e06f787c2a14268595b/servo/components/style/style_adjuster.rs#550